File: android/app/src/main/java/app/notomorrow/app/StoreLoader.kt
package app.notomorrow.app

import android.util.Log
import app.notomorrow.data.db.NoTomorrowDatabase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale
import java.util.concurrent.CancellationException

/**
 * Opens the database, and when it cannot, says so instead of carrying on with an empty stand-in —
 * the port of `StoreLoader.swift`. There is no in-memory fallback any more: a store that silently
 * ran in memory looked like the real one and lost everything logged into it at the next launch.
 *
 * While the store is not open, `RootScreen` shows [StoreErrorScreen] instead of the app, so
 * nothing reads [database]. Nothing is ever deleted: the files are moved aside only when the user
 * picks "Start with empty data" ([moveAside]), into `Store backups/<yyyy-MM-dd HH.mm.ss>/` next to
 * the database. A later build that fixes the cause opens the untouched store as if nothing
 * happened.
 */
class StoreLoader(
    /** `databases/NoTomorrow`; its `-wal`, `-shm` and `-journal` files sit next to it. */
    private val databaseFile: File,
    /** Where [shareableCopies] puts the copies the share sheet reads (`cache/export/store`). */
    private val shareDir: File,
    /** `DatabaseModule.build`: opens the file, runs the migrations, throws when it cannot. */
    private val openDatabase: () -> NoTomorrowDatabase,
    private val clock: () -> Instant = Instant::now,
    private val zone: ZoneId = ZoneId.systemDefault(),
) {

    sealed interface State {
        /** Not tried yet (the splash screen is still up). */
        data object Loading : State

        data object Open : State

        /** [message] is the raw error, for a bug report: not translated, so the screen keeps it collapsed. */
        data class Failed(val message: String) : State
    }

    private val _state = MutableStateFlow<State>(State.Loading)
    val state: StateFlow<State> = _state.asStateFlow()

    val isOpen: Boolean get() = _state.value == State.Open

    /** There is something on disk to share (the error screen hides "Save a copy" otherwise). */
    val hasFiles: Boolean get() = storeFiles(databaseFile).isNotEmpty()

    /** The folder the last [moveAside] filled, for the log and for a later restore. */
    var lastBackup: File? = null
        private set

    @Volatile
    private var opened: NoTomorrowDatabase? = null

    private val lock = Mutex()

    /**
     * The open database. The UI is gated on [state], so nothing asks while it is failed; if
     * something does, it gets an exception rather than a throwaway database that would swallow
     * whatever it writes.
     */
    fun database(): NoTomorrowDatabase =
        opened ?: throw IllegalStateException("The store is not open: ${_state.value}")

    /** Opens the store (Try again calls it again). `true` once it is open; the state says why not. */
    suspend fun open(): Boolean = lock.withLock {
        if (opened != null) return@withLock true
        try {
            opened = withContext(Dispatchers.IO) { openDatabase() }
            _state.value = State.Open
            true
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            Log.e(TAG, "The store could not be opened", e)
            _state.value = State.Failed(e.toString())
            false
        }
    }

    /**
     * "Start with empty data": moves the files that cannot be opened into a new backup folder next
     * to the database, so the next [open] creates an empty one. Never deletes anything; `false`
     * (state unchanged) when a file could not be moved.
     */
    suspend fun moveAside(): Boolean = lock.withLock {
        if (opened != null) return@withLock false
        try {
            lastBackup = withContext(Dispatchers.IO) {
                moveAside(databaseFile, backupsDir(databaseFile), backupStamp(clock(), zone))
            }
            true
        } catch (e: CancellationException) {
            throw e
        } catch (e: IOException) {
            Log.e(TAG, "The store files could not be moved aside", e)
            false
        }
    }

    /**
     * "Save a copy of the data": fresh copies of the store files in [shareDir] (the share sheet
     * reads them through the `FileProvider`), so what is shared is a snapshot and the originals
     * stay untouched. Empty when there is nothing to copy.
     */
    suspend fun shareableCopies(): List<File> = withContext(Dispatchers.IO) {
        runCatching {
            shareDir.deleteRecursively()
            shareDir.mkdirs()
            storeFiles(databaseFile).map { source ->
                File(shareDir, source.name).also { copy ->
                    Files.copy(source.toPath(), copy.toPath(), StandardCopyOption.REPLACE_EXISTING)
                }
            }
        }.getOrElse { emptyList() }
    }

    companion object {
        private const val TAG = "StoreLoader"

        /** The folder every backup goes in, next to the database. */
        const val BACKUPS_FOLDER = "Store backups"

        private val stampFormat = DateTimeFormatter.ofPattern("yyyy-MM-dd HH.mm.ss", Locale.ROOT)

        fun backupsDir(databaseFile: File): File = File(databaseFile.absoluteFile.parentFile, BACKUPS_FOLDER)

        /** `yyyy-MM-dd HH.mm.ss`: sorts by time and is safe as a folder name. */
        fun backupStamp(at: Instant, zone: ZoneId): String = stampFormat.format(at.atZone(zone))

        /** The database file and its SQLite companions that exist on disk, the database first. */
        fun storeFiles(databaseFile: File): List<File> =
            listOf("", "-wal", "-shm", "-journal")
                .map { File(databaseFile.path + it) }
                .filter { it.isFile }

        /**
         * Moves every file of [databaseFile] into `backups/<stamp>` (`<stamp> 2`, `<stamp> 3`… when
         * taken) and returns that folder. Throws before anything moves when the folder cannot be
         * made; a failed move leaves the rest where they are.
         */
        fun moveAside(databaseFile: File, backups: File, stamp: String): File {
            var target = File(backups, stamp)
            var n = 2
            while (target.exists()) target = File(backups, "$stamp ${n++}")
            if (!target.mkdirs()) throw IOException("Could not create $target")
            for (file in storeFiles(databaseFile)) {
                Files.move(file.toPath(), File(target, file.name).toPath())
            }
            return target
        }
    }
}

File: NoTomorrow/App/StoreLoader.swift
import Foundation
import Observation
import SwiftData

/// Opens the SwiftData store and holds the container the app runs on.
/// A store that fails to open is never swapped for an empty one behind the user's back (an in-memory fallback looks
/// like the real app and loses everything logged into it on the next launch). `failure` is set instead and RootView
/// shows `StoreErrorView`: try again, save a copy of the files, or start fresh, which first moves every store file
/// into a timestamped backup folder next to the store (`moveAside`) and then opens a new one (`retry`). Nothing is ever deleted, and a later build that fixes the
/// cause opens the untouched store as if nothing had happened.
@Observable
@MainActor
final class StoreLoader {
    /// The open store, or an empty in-memory placeholder while `failure` is set (no screen reads it then).
    private(set) var container: ModelContainer
    /// Why the store did not open (the raw error, shown in small print); nil while it is open.
    private(set) var failure: String?
    /// Bumped every time a store opens, so the launch work (library import, seeding) reruns after a recovery.
    private(set) var generation = 0
    /// Where the last "start fresh" moved the old files.
    private(set) var lastBackup: URL?

    let configuration: ModelConfiguration
    private let schema: Schema
    private let fileManager: FileManager

    /// Folder next to the store that "start fresh" moves the old files into, one timestamped subfolder per move.
    static let backupFolderName = "Store backups"

    init(schema: Schema = Schema(NoTomorrowSchema.models), configuration: ModelConfiguration? = nil,
         fileManager: FileManager = .default) {
        let configuration = configuration ?? ModelConfiguration("NoTomorrow", schema: schema, isStoredInMemoryOnly: false)
        self.schema = schema
        self.configuration = configuration
        self.fileManager = fileManager
        do {
            container = try ModelContainer(for: schema, configurations: [configuration])
        } catch {
            container = Self.placeholder(schema)
            failure = String(describing: error)
        }
    }

    var isOpen: Bool { failure == nil }

    /// Tries to open the store again (a full disk or a locked device can clear up). True when it opened.
    @discardableResult
    func retry() -> Bool {
        do {
            container = try ModelContainer(for: schema, configurations: [configuration])
            failure = nil
            generation += 1
            return true
        } catch {
            failure = String(describing: error)
            return false
        }
    }

    /// First half of "start fresh": moves the store's files into `Store backups/<timestamp>/` (never deletes them).
    /// True when nothing of the old store is left in place. The caller then resets whatever belonged to the old data
    /// (`hasOnboarded`) before `retry()` opens the new, empty store, so a reopen that fails can't leave the app pointing
    /// at data that is gone. Refused while the store is open.
    @discardableResult
    func moveAside(now: Date = .now) -> Bool {
        guard !isOpen else { return false }
        let files = storeFiles(includingSupport: true)
        guard !files.isEmpty else { return true }
        let stamp = Self.backupStamp(now)
        var backup = backupsFolder.appendingPathComponent(stamp, isDirectory: true)
        var n = 2
        while fileManager.fileExists(atPath: backup.path) {
            backup = backupsFolder.appendingPathComponent("\(stamp) \(n)", isDirectory: true)
            n += 1
        }
        do {
            try fileManager.createDirectory(at: backup, withIntermediateDirectories: true)
            for file in files {
                try fileManager.moveItem(at: file, to: backup.appendingPathComponent(file.lastPathComponent))
            }
            lastBackup = backup
            return true
        } catch {
            // Whatever did move is in `backup`; the rest stays where it was.
            if fileManager.fileExists(atPath: backup.path) { lastBackup = backup }
            failure = String(describing: error)
            return false
        }
    }

    // MARK: Files

    /// The SQLite file and its `-wal` / `-shm` companions, for "save a copy".
    var shareableFiles: [URL] { storeFiles(includingSupport: false) }

    var backupsFolder: URL {
        configuration.url.deletingLastPathComponent().appendingPathComponent(Self.backupFolderName, isDirectory: true)
    }

    /// Every file that belongs to the store: `NoTomorrow.store`, `-wal`, `-shm`, plus Core Data's hidden
    /// `.NoTomorrow_SUPPORT` folder (external blobs) when asked and present.
    func storeFiles(includingSupport: Bool) -> [URL] {
        let url = configuration.url
        let folder = url.deletingLastPathComponent()
        let name = url.lastPathComponent
        let support = "." + url.deletingPathExtension().lastPathComponent + "_SUPPORT"
        let items = (try? fileManager.contentsOfDirectory(at: folder, includingPropertiesForKeys: nil)) ?? []
        return items
            .filter { item in
                let n = item.lastPathComponent
                return n == name || n.hasPrefix(name + "-") || (includingSupport && n == support)
            }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }
    }

    /// "2026-09-22 10.15.00": sortable, and safe as a folder name.
    static func backupStamp(_ date: Date) -> String {
        let c = Calendar(identifier: .gregorian).dateComponents(in: .current, from: date)
        return String(format: "%04d-%02d-%02d %02d.%02d.%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0,
                      c.hour ?? 0, c.minute ?? 0, c.second ?? 0)
    }

    private static func placeholder(_ schema: Schema) -> ModelContainer {
        // An in-memory store with this schema cannot fail short of the process being out of memory.
        try! ModelContainer(for: schema, configurations: [ModelConfiguration(isStoredInMemoryOnly: true)])
    }
}
